### ISSUE TITLE
Errors in the log when HA policy changes in a short period of time

### ISSUE THREAD

**Issue by smee** [rabbitmq/rabbitmq-server#372]
I'm currently experimenting with HA via policies. I see a strange behaviour in the following case:

- RabbitMQ 3.5.5, Windows 7 64bit, Erlang R16B01 32bit
- activated plugins: rabbitmq_event_exchange,rabbitmq_federation,rabbitmq_federation_management,rabbitmq_management,rabbitmq_management_visualiser,rabbitmq_prio_exchange,rabbitmq_shovel,rabbitmq_shovel_management,rabbitmq_top,rabbitmq_tracing
- local cluster, two nodes, empty state
- configure one vhost, one queue, default guest user
- start one producer, one consumer writing/reading to/from the only queue, sleeping between messages
- in parallel to the last step: every 500ms add or remove a policy with attributes `"ha-mode":"all"` and `"ha-sync-mode":"automatic"` matching all queues

As a result I get a number of stacktraces in the log files of both nodes. Also, in our application currently under development we then see inconsistent states in the UI: Some queues still show the name of policy no longer present, some queues are not synchronized with the other node, a shovel that reads from a HA queue stays in state "starting".

I uploaded the complete [archive with mnesia db and logs](http://www.informatik.uni-leipzig.de/~sdienst/rabbits-racecondition-ha-queue.tgz) because the error messages are too long.

**Comment by smee** [comment]
There is no problem in the single node case, only in clusters.

**Comment by michaelklishin** [comment]
Please try with `3.5.6` on 64 bit Erlang and re-upload the results.

**Comment by michaelklishin** [comment]
There are two process crash reports that are known:

```
=CRASH REPORT==== 20-Oct-2015::16:11:20 ===
  crasher:
    initial call: gen:init_it/6
    pid: <0.1023.0>
    registered_name: []
    exception exit: {{badmatch,{error,{"d:/Dropbox/workspaces/customer-repository/routing-compiler-os/dev/rabbits/db/node2-mnesia/queues/7YGRMMI7HBTA5ZFF529N6D48J/journal.jif",
                                       eacces}}},
                     [{rabbit_queue_index,reset_state,1,[]},
                      {rabbit_variable_queue,reset_qi_state,1,[]},
                      {rabbit_variable_queue,purge_and_index_reset,1,[]},
                      {rabbit_variable_queue,delete_and_terminate,2,[]},
                      {rabbit_priority_queue,delete_and_terminate,2,[]},
                      {gen_server2,terminate,3,[]},
                      {proc_lib,init_p_do_apply,3,
                                [{file,"proc_lib.erl"},{line,239}]}]}
      in function  gen_server2:terminate/3 
    ancestors: [<0.1022.0>,rabbit_amqqueue_sup_sup,rabbit_sup,<0.756.0>]
    messages: []
    links: [<0.1022.0>,<0.1024.0>,#Port<0.22555>]
```

is https://github.com/rabbitmq/rabbitmq-server/issues/341 and is Windows-specific, fixed in `3.5.6`.

The other one is https://github.com/rabbitmq/rabbitmq-server/issues/267. I see no other errors and SASL logs seem to be empty.

Closing as a duplicate of https://github.com/rabbitmq/rabbitmq-server/issues/267.

**Comment by smee** [comment]
Ok, so the problem is known, thanks. As requested, here are the results using Erlang 18.1 64bit and RabbitMQ 3.5.6: [Archive](http://www.informatik.uni-leipzig.de/~sdienst/rabbits-ha-bug-3.5.6-erlang18.1_64.tgz)
Indeed, the exceptions are the same as in #267.

**Comment by michaelklishin** [comment]
@smee thank you, your reports are informative and easy to work with, as usual :+1:

**Comment by smee** [comment]
@michaelklishin Thank you! I've never seen such low latencies in any open source related communication as when interacting with you/the RMQ team. Keep up the good work :smile:
